src/D2L.CodeStyle.Analyzers/Immutability/ReadOnlyParameterAnalyzer.cs

```diff
@@ -0,0 +1,124 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class ReadOnlyParameterAnalyzer : DiagnosticAnalyzer {
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.ReadOnlyParameterIsnt
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalysis );
+		}
+
+		private static void RegisterAnalysis( CompilationStartAnalysisContext context ) {
+
+			Compilation compilation = context.Compilation;
+
+			INamedTypeSymbol readOnlyAttribute = compilation.GetTypeByMetadataName( "D2L.CodeStyle.Annotations.ReadOnlyAttribute" );
+
+			if( readOnlyAttribute == null ) {
+				return;
+			}
+
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeParameter(
+					ctx: ctx,
+					readOnlyAttribute: readOnlyAttribute,
+					syntax: ctx.Node as ParameterSyntax
+				),
+				SyntaxKind.Parameter
+			);
+		}
+
+		private static void AnalyzeParameter(
+			SyntaxNodeAnalysisContext ctx,
+			INamedTypeSymbol readOnlyAttribute,
+			ParameterSyntax syntax
+		) {
+			SemanticModel model = ctx.SemanticModel;
+
+			IParameterSymbol parameter = model.GetDeclaredSymbol( syntax, ctx.CancellationToken );
+
+			if( !IsMarkedReadOnly( readOnlyAttribute, parameter ) ) {
+				return;
+			}
+
+			if( parameter.RefKind != RefKind.None ) {
+				ctx.ReportDiagnostic( Diagnostic.Create(
+					Diagnostics.ReadOnlyParameterIsnt,
+					syntax.GetLocation(),
+					"is an in/ref/out parameter"
+				) );
+			}
+
+			IMethodSymbol method = parameter.ContainingSymbol as IMethodSymbol;
+			BlockSyntax methodBody = ( method.DeclaringSyntaxReferences[0].GetSyntax( ctx.CancellationToken ) as MethodDeclarationSyntax ).Body;
+
+			DataFlowAnalysis dataflow = model.AnalyzeDataFlow( methodBody );
+			if( dataflow.WrittenInside.Contains( parameter ) ) {
+				ctx.ReportDiagnostic( Diagnostic.Create(
+					Diagnostics.ReadOnlyParameterIsnt,
+					syntax.GetLocation(),
+					"is assigned to and/or passed by reference"
+				) );
+			}
+
+			if( method.ReturnsByRef ) {
+				foreach( var returnStatement in methodBody.DescendantNodes().OfType<ReturnStatementSyntax>() ) {
+					ISymbol returnedThing = model.GetSymbolInfo( returnStatement.Expression ).Symbol;
+
+					if( returnedThing == null ) {
+						continue;
+					}
+
+					if( returnedThing.OriginalDefinition != parameter ) {
+						continue;
+					}
+
+					ctx.ReportDiagnostic( Diagnostic.Create(
+						Diagnostics.ReadOnlyParameterIsnt,
+						syntax.GetLocation(),

```

Comment:
hmm, this one would be reasonable to switch to `returnStatement.GetLocation()` (though perhaps the wording should change then?). For the dataflow, I didn't want to try to find where the write came from.